### PR TITLE
Storing Listens in Postgres DB

### PR DESCRIPTION
**Pull request by pinkeshbadjatiya**
Migrate the code to submit listens to Postgres Database.
Some major changes made in the commits:
- Use `postgresql` and `cassandra` both for storing listens. (Just need to uncomment a line to switch b/w the two)
- Use `sqlalchemy` as DB Driver for postgresql commits.
- Single config file, `config.py` for both ListenBrainz & ListenStore. (Was mentioned as TODO inside code.)
- Time batch inserts in postgresql, and log those who exceed the threshold
   -  `PG_QUERY_TIMEOUT` denotes the threshold, in miliseconds, for listen commits
   -  Default threshold: **10000ms**
- Remove listens from Postgres which are older than a particular time.
   -  Uses `apscheduler`
   -  Runs jobs every `24 hours`.
   -  `MAX_POSTGRES_LISTEN_HISTORY` denotes the time-range, in days, for listens that must be kept. Rest of the listens will be deleted. Default value is **90 days**
- `PG_ASYNC_LISTEN_COMMIT` denotes whether the listens must be commited asynchronously. Default value is **False**
- Add `listenstore-kafka-to-postgres-writer.py` to save listens to postgres. Older script for Cassandra also works fine, provided schema for cassandra tables is appropriate.
- Support showing listens from both `Cassandra` and `Postgres`. (Just need to swap a line.)
- Fix [Issue #210 on sentry](https://sentry.metabrainz.org/metabrainz/lb-production/issues/210/)
   - **Issue:** Python does not support direct string() to int() conversion. [Read More](https://github.com/metabrainz/listenbrainz-server/pull/80/files#r64879400)

### CODE REVIEW

File: admin/sql/create_tables.sql [hunk 1]
```diff
@@ -8,4 +8,18 @@ CREATE TABLE "user" (
...
+  uid             VARCHAR NOT NULL,
+  year            INT,
+  month           INT,
+  day             INT,
+  id              INT NOT NULL,
+  artist_msid     UUID,
+  album_msid      UUID,
+  recording_msid  UUID,
+  json            VARCHAR,
+  PRIMARY KEY(uid, id)
```
Comment (alastair) [inline admin/sql/create_tables.sql]: this should be in the `create_primary_keys.sql` file

File: listenstore/listenstore/kafkaconsumer.py [hunk 2]
```diff
@@ -37,6 +37,7 @@ def start(self, group_name, topic_name):
                 t0 = time()
 
             messages = self.consumer.get_messages(count=CASSANDRA_BATCH_SIZE, block=True, timeout=KAFKA_READ_TIMEOUT)
+            #print 'New batch'
```
Comment (alastair) [inline listenstore/listenstore/kafkaconsumer.py]: whoops. try not to add print messages, especially if they're commented out

File: listenstore/listenstore/listenstore.py [hunk 3]
```diff
@@ -207,8 +166,9 @@ def fetch_listens(self, uid, from_id=None, to_id=None, limit=None):
                     return
 
     def convert_row(self, row):
-        return Listen(data=ujson.loads(row.json), uid=row.uid, timestamp=row.id, album_msid=row.album_msid,
-                      artist_msid=row.artist_msid, recording_msid=row.recording_msid)
+        print(row)
```
Comment (alastair) [inline listenstore/listenstore/listenstore.py]: print!

File: listenstore/listenstore/listenstore.py [hunk 4]
```diff
@@ -221,7 +181,7 @@ def fetch_listens_for_range(self, uid, date_range, from_id, to_id, order, limit=
         query = """SELECT * FROM listens WHERE uid = %(uid)s AND """ + \
                 range_keys(len(date_range)) + \
                 """ AND id > %(from_id)s AND id < %(to_id)s
-                   ORDER BY id """ + ORDER_TEXT[order] + """ LIMIT %(limit)s"""
+                   ORDER BY id """ + ORDER_TEXT[order] + """ LIMIT %(limit)s;"""
```
Comment (alastair) [inline listenstore/listenstore/listenstore.py]: I'm pretty sure you don't need a `;` here

File: listenstore/listenstore/listenstore.py [hunk 5]
```diff
@@ -258,6 +218,98 @@ def fetch_listens_for_range(self, uid, date_range, from_id, to_id, order, limit=
             else:
                 return
 
+class PostgresListenStore(ListenStore):
+    def __init__(self, conf):
+        ListenStore.__init__(self, conf)
+        self.postgre_connection = None
+        self.postgre_cursor = None
+        
+        self.postgres_conn = psycopg2.connect("dbname='' user='listenbrainz'")
```
Comment (alastair) [inline listenstore/listenstore/listenstore.py]: as I said in irc, make sure that you switch to sqlalchemy. Configuration should be in `config.py`, not in the file

File: listenstore/listenstore/kafkaconsumer.py (hunk 2 above)
Comment (pinkeshbadjatiya) [inline listenstore/listenstore/kafkaconsumer.py]: was testing a lot of stuff. Will remove it surely. :)

File: admin/sql/create_tables.sql (hunk 1 above)
Comment (mayhem) [inline admin/sql/create_tables.sql]: Also, it is customary to have the id column for joining purposes:

    id          SERIAL,

Please add that at the top of the table definition.

Next, we need to add more indexes. We'll need an index on uid and possibly on all of the _msid columns. Alastair?

File: listenstore/listenstore/listenstore.py (hunk 5 above)
Comment (pinkeshbadjatiya) [inline listenstore/listenstore/listenstore.py]: Older configurations were hardcoded in listenstore which was not good. It was mentioned as TODO in code. :)
Added support for separate configuration, `listenstore.conf`, for ListenStore.
https://github.com/pinkeshbadjatiya/listenbrainz-server/blob/b1c922ab1c6ba7417160bec7411012df9769948c/listenstore/listenstore.conf.sample

File: admin/sql/create_tables.sql (hunk 1 above)
Comment (pinkeshbadjatiya) [inline admin/sql/create_tables.sql]: - merged `day`, `month` and `year` field into `date` field
- added `id` field

File: admin/sql/create_tables.sql [hunk 6]
```diff
@@ -8,4 +8,15 @@ CREATE TABLE "user" (
 );
 ALTER TABLE "user" ADD CONSTRAINT user_musicbrainz_id_key UNIQUE (musicbrainz_id);
 
+CREATE TABLE listens (
+  id              SERIAL,
+  uid             VARCHAR NOT NULL,
+  date            TIMESTAMP,
```
Comment (mayhem) [inline admin/sql/create_tables.sql]: TIMESTAMP WITH TIME ZONE, please.

File: admin/sql/create_tables.sql [hunk 7]
```diff
@@ -8,4 +8,15 @@ CREATE TABLE "user" (
...
+  timestamp       INT NOT NULL,
```
Comment (mayhem) [inline admin/sql/create_tables.sql]: Why have two? you don't need a date field if you have a timestamp -- the timestamp includes the date. Please remove the date field and change TIMESTAMP to TIMESTAMP WITH TIME ZONE.

File: admin/sql/create_tables.sql [hunk 8]
```diff
@@ -8,4 +8,15 @@ CREATE TABLE "user" (
...
+  artist_msid     UUID,
+  album_msid      UUID,
+  recording_msid  UUID,
+  json            VARCHAR
```
Comment (mayhem) [inline admin/sql/create_tables.sql]: I think all the UUIDs should be NOT NULL. Please confirm @alastair.

File: admin/sql/create_tables.sql (hunk 7 above)
Comment (pinkeshbadjatiya) [inline admin/sql/create_tables.sql]: My idea was,
`date` - The time when the listen was submitted
`timestamp` - Timestamp of the listen.

Comment (mayhem) [inline admin/sql/create_tables.sql]: What is the use case for this? When will we ever need to know when it was submitted?

Comment (pinkeshbadjatiya) [inline admin/sql/create_tables.sql]: I tried to keep maximal information about the users activity. Did not think much.
I'll remove it. :)

File: admin/sql/create_tables.sql (hunk 8 above)
Comment (pinkeshbadjatiya) [inline admin/sql/create_tables.sql]: I was not sure about the messybrainz DB and what would happen in case listen not indexed.

Comment (alastair) [inline admin/sql/create_tables.sql]: our minimum required text data is artist + track name, right? so perhaps those 2 could be not null. album id could be null

Comment (mayhem) [inline admin/sql/create_tables.sql]: Ah yes, correct. @pinkeshbadjatiya: Please make the artist and recording_msid NOT NULL.

If it isn't indexed, it will still create an artist msid and rcording_msid, so those should be required.

File: listenstore/listenstore/listenstore.py [hunk 9]
```diff
@@ -104,21 +110,126 @@ class ListenStore(object):
...
+        return self.fetch_listens_from_database(uid, from_id, to_id, limit, order, precision)
+
+
+class PostgresListenStore(ListenStore):
+    def __init__(self, conf):
+        ListenStore.__init__(self, conf)
+        self.log.info('Connecting to postgresql: %s', conf['sqlalchemy_database_uri'])
+        self.engine = create_engine(conf['sqlalchemy_database_uri'], poolclass=NullPool)
+        self.connection = self.engine.connect()
+        self.connection.execute("SET synchronous_commit TO off;")
```
Comment (mayhem) [inline listenstore/listenstore/listenstore.py]: What does this do?

File: listenstore/listenstore/listenstore.py [hunk 10]
```diff
@@ -104,21 +110,126 @@ class ListenStore(object):
...
+                VALUES ( %(uid)s, to_timestamp(%(id)s), %(artist_msid)s, %(album_msid)s,
+                %(recording_msid)s, %(json)s) ON CONFLICT DO NOTHING """, values)
+        except sqlalchemy.exc.DataError, e:     # Database error
+            if not self.connection.closed:
+                self.connection.close()
+            print(e)
+
+    @timeout(5)
+    def insert_batch(self, listens):
+        """ Insert a batch of listens, using asynchronous queries.
```
Comment (mayhem) [inline listenstore/listenstore/listenstore.py]: In this case we should have only one function, which is the insert_batch function, which takes an array of listens. If you want to insert only one listen, send an array with only one listen.
Next, you need to add transaction support to the insert function. See https://gist.github.com/alastair/1bbdf4284d0de05de6cb5f5afdeda2f1 for how to do this.

If you do not use a transaction, then it will implicitly use a transaction for you, which is really slow. Opening a transaction at the beginning, insert row(s) and then commit the transaction is the way to go here.

File: admin/sql/create_tables.sql (hunk 8 above)
Comment (pinkeshbadjatiya) [inline admin/sql/create_tables.sql]: done.

File: admin/sql/create_tables.sql (hunk 7 above)
Comment (pinkeshbadjatiya) [inline admin/sql/create_tables.sql]: done.

File: listenstore/listenstore/listenstore.py (hunk 10 above)
Comment (pinkeshbadjatiya) [inline listenstore/listenstore/listenstore.py]: done.

File: listenstore/listenstore/listenstore.py (hunk 9 above)
Comment (pinkeshbadjatiya) [inline listenstore/listenstore/listenstore.py]: Asynchronous commit is an option that allows transactions to complete(virtually complete) more quickly. Since the no of commits are quite large in each transaction, and the length of each transaction is about 300 listens, so making the commit process async may boost up the actual speed at which the lastfm receives the listens.

I was not sure how to enable this(other than manually changing the postgres config file), but my guess was to switch it on before every transaction.

I really think this might help get a speed up. More details.
https://www.postgresql.org/docs/9.4/static/wal-async-commit.html

File: admin/sql/create_tables.sql [hunk 11]
```diff
@@ -8,4 +8,14 @@ CREATE TABLE "user" (
 );
 ALTER TABLE "user" ADD CONSTRAINT user_musicbrainz_id_key UNIQUE (musicbrainz_id);
 
+CREATE TABLE listens (
+  id              SERIAL,
+  uid             VARCHAR NOT NULL,
```
Comment (gentlecat) [inline admin/sql/create_tables.sql]: What's `uid`?

File: admin/sql/create_tables.sql [hunk 12]
```diff
@@ -8,4 +8,14 @@ CREATE TABLE "user" (
...
+  ts              TIMESTAMP WITH TIME ZONE NOT NULL,
+  artist_msid     UUID NOT NULL,
+  album_msid      UUID,
+  recording_msid  UUID NOT NULL,
+  json            VARCHAR
```
Comment (gentlecat) [inline admin/sql/create_tables.sql]: There must be a better name for this column, no?

File: listenstore/bin/listenstore-kafka-to-cassandra-writer.py [hunk 13]
```diff
@@ -13,6 +15,12 @@ def __init__(self):
     def run(self):
         self.kafkaConsumer.start_listens(self.listenStore)
 
+    @property
+    def listenStore(self):
```
Comment (gentlecat) [inline listenstore/bin/listenstore-kafka-to-cassandra-writer.py]: `listen_store` http://pep8.org/#method-names-and-instance-variables

File: listenstore/bin/listenstore-kafka-to-postgres-writer.py [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+
+    def __init__(self):
+        super(KafkaToPostgres, self).__init__()
+        self.log = logging.getLogger(__name__)
+
+    def run(self):
+        self.kafkaConsumer.start_listens(self.listenStore)
+
+    @property
+    def listenStore(self):
```
Comment (gentlecat) [inline listenstore/bin/listenstore-kafka-to-postgres-writer.py]: [Same](https://github.com/metabrainz/listenbrainz-server/pull/80/files#r64876205) here.

File: listenstore/listenstore/listenstore.py [hunk 15]
```diff
@@ -104,21 +110,117 @@ class ListenStore(object):
...
+            order = ORDER_ASC
+        else:
+            order = ORDER_DESC
+
+        precision = 'month'
+        if from_id is None:
+            from_id = MIN_ID
+        if to_id is None:
+            to_id = self.max_id()
+        return self.fetch_listens_from_database(uid, from_id, to_id, limit, order, precision)
```
Comment (gentlecat) [inline listenstore/listenstore/listenstore.py]: I don't see `fetch_listens_from_database` method in this class. It might be better to define it here and raise `NotImplementedError`.

Comment (gentlecat) [inline listenstore/listenstore/listenstore.py]: Also `fetch_listens_from_storage` or something like that would be a more appropriate name, since you don't know where the actual implementation will store listens. It might be in memory or a file instead of database.

File: listenstore/listenstore/listenstore.py [hunk 16]
```diff
@@ -104,21 +110,117 @@ class ListenStore(object):
...
         self.log = logging.getLogger(__name__)
-        self.replication_factor = conf.get("replication_factor", 1)
-        self.keyspace = conf.get("cassandra_keyspace", "listenbrainz")
-        host = conf.get("cassandra_server", "localhost:9092")
+
+    def max_id(self):
+        return int(self.MAX_FUTURE_SECONDS + calendar.timegm(time.gmtime()))
+
+    def fetch_listens(self, uid, from_id=None, to_id=None, limit=None):
+        """ Setup from_id, to_id, and limit for fetching listens
```
Comment (gentlecat) [inline listenstore/listenstore/listenstore.py]: Does this method save `from_id`, `to_id`, and `limit` parameters for fetching listens in the future with some other method?

File: admin/sql/create_tables.sql (hunk 11 above)
Comment (pinkeshbadjatiya) [inline admin/sql/create_tables.sql]: username. (the one we use to login)

File: listenstore/listenstore/listenstore.py [hunk 17]
```diff
@@ -104,21 +110,117 @@ class ListenStore(object):
...
+            to_id = self.max_id()
+        return self.fetch_listens_from_database(uid, from_id, to_id, limit, order, precision)
+
+
+class PostgresListenStore(ListenStore):
+    def __init__(self, conf):
+        ListenStore.__init__(self, conf)
+        self.log.info('Connecting to postgresql: %s', conf['sqlalchemy_database_uri'])
+        self.engine = create_engine(conf['sqlalchemy_database_uri'], poolclass=NullPool)
+        #self.connection.execute("SET synchronous_commit TO off;")
```
Comment (gentlecat) [inline listenstore/listenstore/listenstore.py]: You shouldn't keep commented-out code. Just remove it, or at least add a `TODO` comment explaining why it's here.

File: listenstore/listenstore/listenstore.py [hunk 18]
```diff
@@ -104,21 +110,117 @@ class ListenStore(object):
...
+class CassandraListenStore(ListenStore):
+    def __init__(self, conf):
+        ListenStore.__init__(self, conf)
+        self.replication_factor = conf['cassandra_replication_factor']
+        self.keyspace = conf["cassandra_keyspace"]
+        host = conf["cassandra_server"]
         self.log.info('Connecting to cassandra: %s', host)
         self.cluster = Cluster([host])
-
+
```
Comment (gentlecat) [inline listenstore/listenstore/listenstore.py]: 💣

File: listenstore/listenstore/listenstore.py (hunk 17 above)
Comment (pinkeshbadjatiya) [inline listenstore/listenstore/listenstore.py]: https://github.com/metabrainz/listenbrainz-server/pull/80#discussion_r64810438

File: listenstore/listenstore/utils.py [hunk 19]
```diff
@@ -3,23 +3,32 @@
...
-    return {
-        "kafka_server": "localhost:9092",
-        "cassandra_server": "localhost",
-        "cassandra_keyspace": "listenbrainz"
-    }
+import ConfigParser
+import os
+
+def config(opt_vars):
+    """ Parse config file else return default values """
```
Comment (gentlecat) [inline listenstore/listenstore/utils.py]: In other projects we just tend to use Python files for storing configuration. Do we actually need this? @alastair?

Comment (pinkeshbadjatiya) [inline listenstore/listenstore/utils.py]: We were using builtin flask import feature for importing config files. Since listenstore is not integrated with flask, we can't use that feature.

File: webserver/__init__.py [hunk 20]
```diff
@@ -19,6 +25,9 @@ def create_app():
     import config
     app.config.from_object(config)
 
+    # Run the scheduled jobs
+    scheduledJobs = ScheduledJobs(app.config)
```
Comment (gentlecat) [inline webserver/__init__.py]: I'm not sure if it's a good idea to do this when creating an app.

File: webserver/views/user.py [hunk 21]
```diff
@@ -29,18 +29,20 @@ def lastfmscraper(user_id):
 
 @user_bp.route("/<user_id>")
 def profile(user_id):
-    cassandra = webserver.create_cassandra()
+    # Which database to use to showing user listens.
+    #db_conn = webserver.create_cassandra()
+    db_conn = webserver.create_postgres()
```
Comment (gentlecat) [inline webserver/views/user.py]: Again, commented out code. You can make this an option in config file.

File: listenstore/listenstore/listenstore.py (hunk 9 above)
Comment (mayhem) [inline listenstore/listenstore/listenstore.py]: OK, interesting. Let's keep it as is for now, but make it a configurable option that uses normal commits by default. Then we can play with this once we see the load rising on the server.

File: webserver/views/user.py [hunk 22]
```diff
@@ -29,18 +29,20 @@ def lastfmscraper(user_id):
...
 
     # Getting data for current page
     max_ts = request.args.get("max_ts")
     if max_ts is not None:
         try:
-            max_ts = int(max_ts)
+            max_ts = int(float(max_ts))
```
Comment (gentlecat) [inline webserver/views/user.py]: Why is that needed?

File: admin/sql/create_tables.sql (hunk 12 above)
Comment (mayhem) [inline admin/sql/create_tables.sql]: Maybe data? raw_data? Also, the column type should be jsonb, not VARCHAR.

File: webserver/__init__.py (hunk 20 above)
Comment (pinkeshbadjatiya) [inline webserver/__init__.py]: do you mean "We can start scheduledJobs it a bit later" or "Their must be a better way to do it" ?

File: listenstore/listenstore/listenstore.py (hunk 17 above)
Comment (gentlecat) [inline listenstore/listenstore/listenstore.py]: @mayhem suggested to make in configurable. I agree with that.

File: admin/sql/create_tables.sql (hunk 12 above)
Comment (gentlecat) [inline admin/sql/create_tables.sql]: What do we store there?

File: listenstore/listenstore/utils.py (hunk 19 above)
Comment (mayhem) [inline listenstore/listenstore/utils.py]: In that case, the configuration should be passed in from the main application and the main application can store these bits in python files.

File: webserver/__init__.py (hunk 20 above)
Comment (mayhem) [inline webserver/__init__.py]: Yeah, scheduled jobs should just be started by the cron daemon.

File: admin/sql/create_tables.sql (hunk 11 above)
Comment (gentlecat) [inline admin/sql/create_tables.sql]: Why not `id`? `id` is a PK in `user` table.

Comment (gentlecat) [inline admin/sql/create_tables.sql]: Also, `uid` can be confusing. Better make it `username` or `user_id`.

File: admin/sql/create_tables.sql (hunk 12 above)
Comment (mayhem) [inline admin/sql/create_tables.sql]: The raw data submitted by the user is stored there -- the user is allowed to put other pieces of data into this space that we can't know to stick into the DB. This is all good.

File: webserver/views/user.py (hunk 22 above)
Comment (pinkeshbadjatiya) [inline webserver/views/user.py]: yeah. that is a good bug.
Python does not provide direct `string` to `int` conversion. It was failing when opening https://listenbrainz.org/user/armalcolite?max_ts=1459858239.0 i.e. when the timestamps are ending with decimals. (It throws on existing LB website as well)

On upgrading to postgres, the timestamp conversion ends with a decimal precision, and thus it was a problem. We could either reduce the precision in postgres, or do this to encorporate this change.

File: webserver/__init__.py (hunk 20 above)
Comment (gentlecat) [inline webserver/__init__.py]: I'm saying when I run `create_app` function, I don't expect it to do anything else apart from things directly related to creating an instance of Flask app.

File: listenstore/listenstore/utils.py (hunk 19 above)
Comment (pinkeshbadjatiya) [inline listenstore/listenstore/utils.py]: That is again a problem.
We access Listenstore from 2 places, once while storing DB which is from the LB app, hence we can take into account what you said, but when accessing Listenstore for showing listens, we are not accessing it from the flask app. 

This was the problem i faced, and the workaround i found was to keep a copy of configuration in `listenstore.conf` and `config.py` (of the main LB app)

File: listenstore/listenstore/listenstore.py [hunk 23]
```diff
@@ -104,21 +110,117 @@ class ListenStore(object):
...
+    def format_dict(self, listen):
+        return { 'uid': listen.uid,
+                 'id': listen.timestamp,
+                 'artist_msid': uuid.UUID(listen.artist_msid),
+                 'album_msid': uuid.UUID(listen.album_msid) if listen.album_msid is not None else None,
+                 'recording_msid': uuid.UUID(listen.recording_msid),
+                 'json': ujson.dumps(listen.data)}
+
+    @timeout(5)
+    def insert_postgresql(self, listens):
```
Comment (mayhem) [inline listenstore/listenstore/listenstore.py]: I'm really glad you thought about database timeouts here! However, this mechanism isn't the proper one in this context, since the underlying database supports them. Also, deciding what to do with timeouts is hard to do now. In this case, I would suggest the following:

1. Remove this timeout feature.
2. Add a function to keep track how long postgres spends on the query.
3. Add a configuration option that allows users to specify a threshold at which point slow queries get logged. (e.g. if 3000 milliseconds are specified, then print a log statement for any query that takes longer than threshold (3000ms) to execute. We can then keep our eyes on the logs and see when things start getting into trouble.

File: listenstore/listenstore/listenstore.py [hunk 24]
```diff
@@ -104,21 +110,117 @@ class ListenStore(object):
...
+            for listen in listens:
+                if not listen.validate():
+                    raise ValueError("Invalid listen: %s" % listen)
+                try:
+                    res = connection.execute(
+                    """INSERT INTO listens(uid, ts, artist_msid, album_msid, recording_msid, json)
+                        VALUES ( %(uid)s, to_timestamp(%(id)s), %(artist_msid)s, %(album_msid)s,
+                        %(recording_msid)s, %(json)s) ON CONFLICT DO NOTHING """, self.format_dict(listen))
+                except sqlalchemy.exc.DataError, e:     # Database error
+                    print(e)
```
Comment (mayhem) [inline listenstore/listenstore/listenstore.py]: You should setup proper logging for this application and not use print statements. This applies to all places in the app where you are inclined to print debug/informational stuff.

File: listenstore/listenstore/listenstore.py (hunk 18 above)
Comment (mayhem) [inline listenstore/listenstore/listenstore.py]: Whitespace nazi!

File: admin/sql/create_tables.sql (hunk 12 above)
Comment (pinkeshbadjatiya) [inline admin/sql/create_tables.sql]: done!

File: admin/sql/create_tables.sql (hunk 11 above)
Comment (mayhem) [inline admin/sql/create_tables.sql]: +1 to user_id

File: listenstore/bin/listenstore-kafka-to-cassandra-writer.py (hunk 13 above)
Comment (pinkeshbadjatiya) [inline listenstore/bin/listenstore-kafka-to-cassandra-writer.py]: done!

File: listenstore/bin/listenstore-kafka-to-postgres-writer.py (hunk 14 above)
Comment (pinkeshbadjatiya) [inline listenstore/bin/listenstore-kafka-to-postgres-writer.py]: done!

File: webserver/views/user.py (hunk 21 above)
Comment (pinkeshbadjatiya) [inline webserver/views/user.py]: I am still not sure if i must remove the cassandra support completely. In that i may need to do some code refactoring, initially i tried to make it compatible with both. I think its better to just remove Cassandra code.

@mayhem fine?

File: admin/sql/create_tables.sql (hunk 11 above)
Comment (pinkeshbadjatiya) [inline admin/sql/create_tables.sql]: done!

File: listenstore/listenstore/listenstore.py (hunk 16 above)
Comment (pinkeshbadjatiya) [inline listenstore/listenstore/listenstore.py]: no. these are passed on every request.

Comment (gentlecat) [inline listenstore/listenstore/listenstore.py]: Then this docstring needs an update.

Comment (pinkeshbadjatiya) [inline listenstore/listenstore/listenstore.py]: Setup in the sense, if the user has given valid values or not. Also, if none is given, then use default values and pass it to the next method,  `fetch_listens_from_database()`

File: listenstore/bin/listenstore-kafka-to-cassandra-writer.py (hunk 13 above)
Comment (pinkeshbadjatiya) [inline listenstore/bin/listenstore-kafka-to-cassandra-writer.py]: There's a catch :'(
All the methods which are marked as `@property` follow CamelCase notation in the existing LB structure. (ex. listenStore, kafkaConsumer). This might violate PEP8 standards, but i believe its good to keep it the way it is. Sometimes even i find it hard to track variables and methods, considering so diverse structure of ListenBrainz.

Though, changing it will require considerable change in a lot of method names.

Comment (gentlecat) [inline listenstore/bin/listenstore-kafka-to-cassandra-writer.py]: Well, you can put that in a separate commit if you don't want to mix things together. But that's the convention we use in all of our projects.

**Comment by gentlecat** [comment]
It would also be nice to see some tests for things that you are adding.

File: listenstore/listenstore/utils.py (hunk 19 above)
Comment (gentlecat) [inline listenstore/listenstore/utils.py]: You can load configuration values from config file written in regular Python using this https://github.com/pallets/flask/blob/master/flask/config.py#L111-L137. Can just reuse that class since Flask is one of current requirements.

File: listenstore/bin/listenstore-kafka-to-cassandra-writer.py (hunk 13 above)
Comment (pinkeshbadjatiya) [inline listenstore/bin/listenstore-kafka-to-cassandra-writer.py]: done!

File: listenstore/listenstore/listenstore.py (hunk 15 above)
Comment (pinkeshbadjatiya) [inline listenstore/listenstore/listenstore.py]: done!

File: listenstore/listenstore/listenstore.py (hunk 16 above)
Comment (pinkeshbadjatiya) [inline listenstore/listenstore/listenstore.py]: done!

File: listenstore/listenstore/utils.py (hunk 19 above)
Comment (pinkeshbadjatiya) [inline listenstore/listenstore/utils.py]: **Done!**
(This is the minimal what i could create, [https://github.com/pinkeshbadjatiya/listenbrainz-server/blob/040dd4a20fa69e96b29affa15777451d7caffe85/listenstore/listenstore/loadconfig.py](https://github.com/pinkeshbadjatiya/listenbrainz-server/blob/040dd4a20fa69e96b29affa15777451d7caffe85/listenstore/listenstore/loadconfig.py) )

File: webserver/__init__.py (hunk 20 above)
Comment (pinkeshbadjatiya) [inline webserver/__init__.py]: done!

**Comment by pinkeshbadjatiya** [comment]
Implementation is not yet finalized for every piece of code. It would be difficult to write tests in such a scenario where things keep changing too frequently. 

Also, a lot of stuff is going on in this PR. Its better to keep tests separate. The previous tests also need to be updated. I would be writing tests for a lot of my previous PR's as well once things get shape. :)

File: listenstore/listenstore/listenstore.py (hunk 24 above)
Comment (pinkeshbadjatiya) [inline listenstore/listenstore/listenstore.py]: done!

File: webserver/views/user.py (hunk 21 above)
Comment (mayhem) [inline webserver/views/user.py]: I'm fine removing it. The old code is in git, so we can always dig it back out.

File: listenstore/listenstore/listenstore.py (hunk 23 above)
Comment (pinkeshbadjatiya) [inline listenstore/listenstore/listenstore.py]: done!

File: listenstore/listenstore/listenstore.py (hunk 17 above)
Comment (pinkeshbadjatiya) [inline listenstore/listenstore/listenstore.py]: Done!

**Comment by gentlecat** [comment]
@brainzbot, add to whitelist.
